asserts/snap_asserts.go

```diff
@@ -485,3 +486,122 @@ func assembleValidation(assert assertionBase) (Assertion, error) {
+func (basedcl *BaseDeclaration) PlugRule(interfaceName string) *PlugRule {
+	return basedcl.plugRules[interfaceName]
+}
+
+// SlotRule returns the slot-side rule about the given interface if one was included in the slots stanza of the declaration, otherwise it returns nil.
+func (basedcl *BaseDeclaration) SlotRule(interfaceName string) *SlotRule {
+	return basedcl.slotRules[interfaceName]
+}
+
+// Implement further consistency checks.
+func (basedcl *BaseDeclaration) checkConsistency(db RODatabase, acck *AccountKey) error {
+	// XXX: not signed or stored yet in a db, but being ready for that
+	if !db.IsTrustedAccount(basedcl.AuthorityID()) {
+		return fmt.Errorf("base-declaration assertion for series %s is not signed by a directly trusted authority: %s", basedcl.Series(), basedcl.AuthorityID())
+	}
+	return nil
+}
+
+// sanity
+var _ consistencyChecker = (*BaseDeclaration)(nil)
+
+func assembleBaseDeclaration(assert assertionBase) (Assertion, error) {
+	var plugRules map[string]*PlugRule
+	plugs, err := checkMap(assert.headers, "plugs")
+	if err != nil {
+		return nil, err
+	}
+	if plugs != nil {
+		plugRules = make(map[string]*PlugRule, len(plugs))
+		for iface, rule := range plugs {
+			plugRule, err := compilePlugRule(iface, rule)
+			if err != nil {
+				return nil, err
+			}
+			plugRules[iface] = plugRule
+		}
+	}
+
+	var slotRules map[string]*SlotRule
+	slots, err := checkMap(assert.headers, "slots")
+	if err != nil {
+		return nil, err
+	}
+	if slots != nil {
+		slotRules = make(map[string]*SlotRule, len(slots))
+		for iface, rule := range slots {
+			slotRule, err := compileSlotRule(iface, rule)
+			if err != nil {
+				return nil, err
+			}
+			slotRules[iface] = slotRule
+		}
+	}
+
+	timestamp, err := checkRFC3339Date(assert.headers, "timestamp")
+	if err != nil {
+		return nil, err
+	}
+
+	return &BaseDeclaration{
+		assertionBase: assert,
+		plugRules:     plugRules,
+		slotRules:     slotRules,
+		timestamp:     timestamp,
+	}, nil
+}
+
+var builtinBaseDeclaration *BaseDeclaration
+
+// BuiltinBaseDeclaration exposes the initialized builtin base-declaration assertion. This is used by overlord/assertstate, other code should use assertstate.BaseDeclaration.
+func BuiltinBaseDeclaration() *BaseDeclaration {
+	return builtinBaseDeclaration
+}
+
+// InitBuiltinBaseDeclaration initializes the builtin base-declaration based on headers (or resets it if headers is nil).
+func InitBuiltinBaseDeclaration(headers []byte) error {
+	if headers == nil {
+		builtinBaseDeclaration = nil
+		return nil
+	}
+	trimmed := bytes.TrimSpace(headers)
+	h, err := parseHeaders(trimmed)
+	if err != nil {
+		return err
+	}
+	if h["type"] != "base-declaration" {
+		return fmt.Errorf("the builtin base-declaration headers sport the wrong type")
+	}
+	h["timestamp"] = time.Now().UTC().Format(time.RFC3339)
+	h["sign-key-sha3-384"] = "lhSOrK6PQ22AJBuQ6btNW2NGAs-lZd4zsd6O5qRY7ylFDQVCD_0rurr66osdF-n9" // sha3-384("$builtin")
```

Comment:
We discussed this a bit on IRC. I was expecting that we would skip any signing checks since the document isn't signed (eg pseudo-code 'if base-declaration, warn else, verify). This code is providing a sha-384 for a non-existent public key which is non-obvious (to me anyway) why that should work and is ok.

That said, I understand that we will be signing the base declaration soonish, so I don't want to waste people's time or to introduce corner cases that might introduce verification bugs, so if @pedronis and @niemeyer as experts in the assertions code are ok with this as is, +1 so long as we add a 'FIXME' comment that details why this is working correctly and is safe. @pedronis thought he might also explore a code change, which should be reviewed by @niemeyer.